File: cpp/0394-decode-string.cpp
#include <stack>
#include <string>

class Solution {
    public:
        [[nodiscard]] auto decodeString(const std::string &s) -> std::string {
            auto curr_s = std::string{};
            curr_s.reserve(s.size());
            auto curr_n = 0U;
            auto s_stack = std::stack<std::string>{};
            auto n_stack = std::stack<unsigned>{};
            for (const auto c : s) {
                if ('a' <= c && c <= 'z') {
                    curr_s += c;
                } else if ('0' <= c && c <= '9') {
                    curr_n = 10 * curr_n + static_cast<unsigned>(c) - '0';
                } else if (c == '[') {
                    s_stack.push(curr_s);
                    n_stack.push(curr_n);
                    curr_s.clear();
                    curr_n = 0;
                } else {
                    // c is "]"
                    const auto n = n_stack.top();
                    n_stack.pop();
                    auto &prev_s = s_stack.top();
                    prev_s.reserve(prev_s.size() + n * curr_s.size());
                    for (auto i = 0U; i < n; ++i) {
                        prev_s.append(curr_s);
                    }
                    curr_s = prev_s;
                    s_stack.pop();
                }
            }

            return curr_s;
        }
};

File: py/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        curr_s = ""
        curr_n = 0
        stack: list[int | str] = []
        for c in s:
            if c.isalpha():
                curr_s += c
            elif c.isdigit():
                curr_n = 10 * curr_n + int(c)
            elif c == "[":
                stack.append(curr_s)
                stack.append(curr_n)
                curr_s = ""
                curr_n = 0
            else:
                # c is "]"
                n = stack.pop()
                assert isinstance(n, int)
                prev_s = stack.pop()
                assert isinstance(prev_s, str)
                curr_s = prev_s + n * curr_s

        return curr_s
